transports/webrtc/src/connection.rs

```diff
@@ -0,0 +1,280 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+mod poll_data_channel;
+
+use futures::{
+    channel::{
+        mpsc,
+        oneshot::{self, Sender},
+    },
+    lock::Mutex as FutMutex,
+    {future::BoxFuture, prelude::*, ready},
+};
+use futures_lite::StreamExt;
+use libp2p_core::{muxing::StreamMuxer, Multiaddr};
+use log::{debug, error, trace};
+use webrtc::data_channel::RTCDataChannel;
+use webrtc::peer_connection::RTCPeerConnection;
+use webrtc_data::data_channel::DataChannel as DetachedDataChannel;
+
+use std::{
+    sync::{Arc, Mutex as StdMutex},
+    task::{Context, Poll},
+};
+
+use crate::error::Error;
+pub(crate) use poll_data_channel::PollDataChannel;
+
+const MAX_DATA_CHANNELS_IN_FLIGHT: usize = 10;
+
+/// A WebRTC connection, wrapping [`RTCPeerConnection`] and implementing [`StreamMuxer`] trait.
+pub struct Connection {
+    /// `RTCPeerConnection` to the remote peer.
+    ///
+    /// Uses futures mutex because used in async code (see poll_outbound and poll_close).
+    peer_conn: Arc<FutMutex<RTCPeerConnection>>,
+
+    inner: StdMutex<ConnectionInner>,
```

Comment:
If I am not mistaken, this `StdMutex` is only needed because the methods on `StreamMuxer` take `&self` and not `&mut self`. Once https://github.com/libp2p/rust-libp2p/pull/2765 is merged, we could thus get rid of it.

Reply:
#2765 is merged now.

Reply:
updated in https://github.com/libp2p/rust-libp2p/pull/2622/commits/59894fbf5c4c9c3375ec6224a8f770386696d16d. @thomaseizinger was I correct to use pin-project or there's an easier way? I'll be honest I'm still new / fuzzy re Pin / Unpin.